src/txmempool.h

```diff
@@ -63,6 +63,64 @@ struct CompareIteratorByHash {
         return a->GetTx().GetHash() < b->GetTx().GetHash();
     }
 };
+
+/** Epoch: RAII-style guard for using epoch-based graph traversal algorithms.
+ *     When walking ancestors or descendants, we generally want to avoid
+ * visiting the same transactions twice. Some traversal algorithms use
+ * std::set (or setEntries) to deduplicate the transaction we visit.
+ * However, use of std::set is algorithmically undesirable because it both
+ * adds an asymptotic factor of O(log n) to traverals cost and triggers O(n)
+ * more dynamic memory allocations.
+ *     In many algorithms we can replace std::set with an internal mempool
+ * counter to track the time (or, "epoch") that we began a traversal, and
+ * check + update a per-transaction epoch for each transaction we look at to
+ * determine if that transaction has not yet been visited during the current
+ * traversal's epoch.
+ *     Algorithms using std::set can be replaced on a one by one basis.
+ * Both techniques are not fundamentally incompatible across the codebase.
+ * Generally speaking, however, the remaining use of std::set for mempool
+ * traversal should be viewed as a TODO for replacement with an epoch based
+ * traversal, rather than a preference for std::set over epochs in that
+ * algorithm.
+ */
+
+class LOCKABLE Epoch
+{
+private:
+    uint64_t m_raw_epoch = 0;
+    bool m_guarded = false;
+
+public:
+    Epoch() { }
+    Epoch(const Epoch&) = delete;
+    Epoch& operator=(const Epoch&) = delete;
+
+    bool guarded() const { return m_guarded; }
+
+    class Marker {
+    private:
+        uint64_t m_marker = 0;
+        friend class Epoch;
+    };
+
+    class SCOPED_LOCKABLE Guard {
+    private:
+        Epoch& m_epoch;
+
+    public:
+        explicit Guard(Epoch& epoch) EXCLUSIVE_LOCK_FUNCTION(epoch);
+        ~Guard() UNLOCK_FUNCTION();
+    };
+
+    bool visited(Marker& marker) const EXCLUSIVE_LOCKS_REQUIRED(*this) {
+        assert(m_guarded);
+        bool ret = marker.m_marker >= m_raw_epoch;
+        if (!ret) marker.m_marker = m_raw_epoch;
+        return ret;
+    }
```

Comment:
That reverses the logic (`ret == false` causes `m_marker` to change originally, but your code returns `true` if `m_marker` changed) ? Seems like that's evidence it's not that easy to read/reason about? :)

It might be less confusing to write it out in full:

```c++
if (marker.m_marker < m_raw_epoch ) {
    // this entry is from an earlier epoch, so it hasn't been visited
    marker.m_marker = m_raw_epoch;
    return false;
} else {
    return true;
}
```

I'd guess the compiler will make all these variants essentially the same anyway, so no objection to any variation from me.

Reply:
Ah my bad, I've made a case against myself -- for what it's worth, I made the proposed version by translating the proposed version, which I have trouble parsing logically, and thought it was equivalent to that logic (and still can't figure out the concrete reason it's opposite).

I'm fine with the completely written out version as it is the most obvious for sleepy brains like mine.

Reply:
I think the translation goes:

- `r = a >= b; if (!r) a = b; return r;`  (current)
- `o = a; if (! (a >= b)) a = b; return o >= b;`   (store old value, replace `r`)
- `o = a; if (a < b) a = b; return !(o < b);`  (switch to less than)
- `o = a; a = max(a,b); return !(o != a);`  (`o < b` iff the if was taken, and `a` was changed)
- `o = a; a = max(a,b); return o == a;`  (simplify)

Anyway, changed to KISS version.